Ble client fixes for proxy

elupus opened the pull request:
# What does this implement/fix?

- Makes sure bluetooth proxy get open failure indication (broken since #5277)
- Makes sure bluetooth proxy detect closes even when there are multiple virtual connections
- Removes unneeded checks from bluetooth proxy (redundant since #5277)

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** Related to #5277 and replaces #6590

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

elupus committed:
Remove checks fro conn_id from proxy

These checks where added to base class in #5277

elupus committed:
Listen to close instead of disconnect in ble proxy

The disconnection event might not occure, if multiple
virtual connections are active to same device

elupus committed:
Make sure ble client users get open errors

The bluetooth proxy need to inform it's controller
of the fact that opening the connection failed.

elupus committed:
Avoid tidy warnings

error: switch has 2 consecutive identical branches [bugprone-branch-clone,-warnings-as-errors]

clydebarrow commented:
This all looks good to me, so I guess the question is what testing has/can be done?

elupus committed:
React to disconnect event too

Unsure if stack will trigger CLOSE events for all
type of disconnections. The samples all listen to
disconnect.

elupus commented:
I have now tested with a "broken" config like below. Where both the proxy through HA connects to the device, as well as the ESP device itself connects to the device. Seems to work now. Previously the proxy would time out it's disconnections since no disconnection even would occure.

Admittedly, that is only a partial test of the changes here thou. The open failure indication i've not been able to figure out a testcase for.

```
substitutions:
  name: esp32-bluetooth-proxy-0b6900
  friendly_name: Bluetooth Proxy 0b6900
packages:
  esphome.bluetooth-proxy: github://esphome/bluetooth-proxies/esp32-generic.yaml@main
esphome:
  name: ${name}
  name_add_mac_suffix: false
  friendly_name: ${friendly_name}
api:
  encryption:
    key: ...

external_components:
  - source: github://pr#6596
    components: [ bluetooth_proxy ]

ble_client:
  - mac_address: F0:5E:CD:31:37:3E
    id:  gardena_water_switch

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
```

clydebarrow reviewed (approved):
All looks good to me.
